components/fxa-client/android/src/main/java/mozilla/appservices/fxa/FirefoxAccount.kt

```diff
@@ -0,0 +1,153 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.fxa
+
+import mozilla.appservices.fxaclient.AccessTokenInfo
+import mozilla.appservices.fxaclient.Device
+import mozilla.appservices.fxaclient.DeviceCapability
+import mozilla.appservices.fxaclient.DevicePushSubscription
+import mozilla.appservices.fxaclient.FxaConfig
+import mozilla.appservices.fxaclient.FirefoxAccount as RustFirefoxAccount
+import mozilla.appservices.fxaclient.IncomingDeviceCommand
+import mozilla.appservices.fxaclient.Profile
+import mozilla.appservices.sync15.DeviceType
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Top-level class to manage a FxA account
+ *
+ * This class contains the top-level functionality like fetching tokens and listening for account
+ * events.  See also:
+ *   - StateManager: manages the account state machine.  Defined in `State.kt` and available from
+ *     the `stateManager()` method.
+ *   - DeviceManager: manages the list of connected devices to an FxA account.  Defined in
+ *     `Devices.kt` and available from the `deviceManager()` method.
+ */
+abstract class FirefoxAccount(
+    private val config: FxaConfig,
+    private val deviceConfig: DeviceConfig,
+    private val storageHandler: StorageHandler,
+    private val applicationScopes: List<Scope> = listOf(),
+) {
+    /**
+     * Tries to fetch an access token for the given scope.
+     *
+     * @param singleScope Single OAuth scope (no spaces) for which the client wants access
+     * @return [AccessTokenInfo] that stores the token, along with its scope, key and
+     *                           expiration timestamp (in seconds) since epoch when complete
+     */
+    abstract suspend fun getAccessToken(singleScope: String): AccessTokenInfo?
+
+    /**
+     * Fetches the profile object for the current client either from the existing cached state
+     * or from the server (requires the client to have access to the profile scope).
+     *
+     * @param ignoreCache Fetch the profile information directly from the server
+     * @return Profile (optional, if successfully retrieved) representing the user's basic profile info
+     */
+    abstract suspend fun getProfile(ignoreCache: Boolean = false): Profile?
+
+    /**
+     * Fetches the token server endpoint, for authentication using the SAML bearer flow.
+     *
+     * @return Token server endpoint URL string, `null` if it couldn't be obtained.
+     */
+    abstract suspend fun getTokenServerEndpointURL(): String?
```

Comment:
should we be using urls in the new api?

Reply:
> If the UniFFI custom converter fails, then I think we would get a panic,

Good point, but I think this is probably solvable in the same way you solved it for callback functions, right? If so, I'll open a uniffi issue